### DISCUSSION TITLE
teaching how to transfer to emulator

### DISCUSSION THREAD

**Discussion by Laranjo1** [vfsfitvnm/frida-il2cpp-bridge#620] [Show and tell]
### First, install `frida-tools`:
```bash
pip install frida-tools
```

After installing `frida-tools`, go to this GitHub page:  
[frida-server-x86](https://github.com/frida/frida/releases/tag/16.5.1)  
Locate and download:  
`frida-server-16.5.1-android-x86.xz`

Then, download this emulator (I’ve only tested with this one):  
[LDPLAYER 5 x86](https://ldcdn.ldmnq.com/download/LDPlayer4.exe?n=LDPlayer4_pt_1280_ld.exe)

After this, you're almost set up!  
Run the following commands:

```bash
npm init -y
npm i -D @types/node @types/frida-gum typescript frida-compile frida-il2cpp-bridge
```

Add this to the `package.json` file generated by `npm init`:

```json
"scripts": {
    "spawn": "frida -U -f com.CrystalDev.DragonCrystalArena --realm emulated -l dist/agent.js",
    "watch": "frida-compile src/index.ts -o dist/agent.js -w",
    "build": "frida-compile src/index.ts -o dist/agent.js -c"
},
```

**Note**: My build script doesn't work, I’m not sure why. Any suggestions would be appreciated!

Next, create the following folder and file structure:

```
src
     └── index.ts
dist
     └── agent.js
```

### Setting up the LDPlayer Emulator:

1. Open the directory where LDPlayer4 is installed.
2. Execute the following commands:

```bash
adb devices
adb push <local path to frida-server-16.5.1-android-x86> "data/local/tmp"
adb shell "chmod 777 data/local/tmp/frida-server-16.5.1-android-x86"
adb shell "data/local/tmp/frida-server-16.5.1-android-x86" &
```

### To run the program:
1. Open a terminal in the project directory and run:
   ```bash
   npm run watch
   ```

2. Open a second terminal and run:
   ```bash
   npm run spawn
   ```

3. Make any change to the code (I usually leave a comment for this).

Ignore the two errors that may appear, and your Frida setup should be operational!

**Comment by namtacs** [comment]
This is the same as following the official guide https://frida.re/docs/android/ and using files from the wiki. The emulator makes no difference in setup from a normal device apart from the need of `--realm emulated` to attach to the emulated ARM realm (libhoudini).

**Reply by doviettung96** [reply]
Does it work now? I still see that people are struggling with --realm emulated and still not succeed to find the lib2ilcpp.so

**Comment by Laranjo1** [comment]
> "And what about Android 9.0 (64-bit)? Can it be used? I tried, but I still couldn't find libil2cpp."

For emulators I think only can x86
